File: app/src/main/java/com/simbirsoft_mobile_revolution2017_v_team/mobilelibrary/ui/fragments/ListBooksFragment/RecyclerViewAdapter.java
package com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.ui.fragments.ListBooksFragment;

import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.R;
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.domain.Book;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Olegka on 04.10.2017.
 */

public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder> {

    public interface OnBookClickListener{
        void onBookClicked(String bookId);
    }

    private List<Book> books;

    private OnBookClickListener adapterClickListener;

    public RecyclerViewAdapter(List<Book> books, OnBookClickListener adapterClickListener) {
        this.books = books;
        this.adapterClickListener = adapterClickListener;
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        public View view;

        @BindView(R.id.name)
        public TextView textName;

        @BindView(R.id.author)
        public TextView textAuthor;

        public String bookId;
        public OnBookClickListener bookClickListener;

        public ViewHolder(View v, OnBookClickListener abstractClickListener) {
            super(v);
            view = v;
            ButterKnife.bind(this, v);
            bookId = "-1";
            bookClickListener = abstractClickListener;
            v.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            bookClickListener.onBookClicked(bookId);
        }
    }

    @Override
    public RecyclerViewAdapter.ViewHolder onCreateViewHolder(ViewGroup parent,
                                                             int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.recycler_item_book, parent, false);

        ViewHolder vh = new ViewHolder(v, adapterClickListener);
        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bookId = books.get(position).getId();
        holder.textName.setText(books.get(position).getName());
        holder.textAuthor.setText(books.get(position).getAuthor());
    }

    @Override
    public int getItemCount() {
        return books.size();
    }
}

File: app/src/main/java/com/simbirsoft_mobile_revolution2017_v_team/mobilelibrary/ui/fragments/ListBooksFragment/RecyclerViewAdapter.kt
package com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.ui.fragments.ListBooksFragment

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.R
import com.simbirsoft_mobile_revolution2017_v_team.mobilelibrary.domain.BookK

/**
 * Created by user on 15.11.2017.
 */
class RecyclerViewAdapterK(
        private val books: List<BookK>,
        private val adapterClickListener: OnBookClickListener
) : RecyclerView.Adapter<RecyclerViewAdapterK.ViewHolder>() {

    interface OnBookClickListener {
        fun onBookClicked(bookId: String)
    }

    inner class ViewHolder(
            view: View,
            private val bookClickListener: OnBookClickListener
    ) : RecyclerView.ViewHolder(view), View.OnClickListener {
        var textName: TextView? = null
        var textAuthor: TextView? = null
        var bookId: String

        init {
            textName = view.findViewById(R.id.name)
            textAuthor = view.findViewById(R.id.author)
            bookId = "-1"
            view.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            bookClickListener.onBookClicked(bookId)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): RecyclerViewAdapterK.ViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.recycler_item_book, parent, false)

        return ViewHolder(v, adapterClickListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bookId = books[position].id!!
        holder.textName?.text = books[position].name
        holder.textAuthor?.text = books[position].author
    }

    override fun getItemCount(): Int = books.size
}